File: internal/xds/snapshot.go
package xds

import (
	"fmt"
	"strings"

	cluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
	endpoint "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
	listener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
	route "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
	router "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/router/v3"
	hcm "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
	tcpproxy "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/tcp_proxy/v3"
	"github.com/envoyproxy/go-control-plane/pkg/cache/types"
	"github.com/envoyproxy/go-control-plane/pkg/cache/v3"
	"github.com/envoyproxy/go-control-plane/pkg/resource/v3"
	"github.com/envoyproxy/go-control-plane/pkg/wellknown"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/durationpb"
)

// BuildSnapshot creates a snapshot with listeners, routes, and clusters
func BuildSnapshot(version string) (*cache.Snapshot, error) {
	// Create HTTP listener on port 80
	httpListener, err := makeHTTPListener()
	if err != nil {
		return nil, fmt.Errorf("failed to create HTTP listener: %w", err)
	}

	// Create empty route configuration (returns 404 for everything)
	routeConfig := makeEmptyRouteConfig()

	// Build snapshot with all resources
	snapshot, err := cache.NewSnapshot(
		version,
		map[resource.Type][]types.Resource{
			resource.ClusterType:  {}, // Empty clusters for now
			resource.RouteType:    {routeConfig},
			resource.ListenerType: {httpListener},
		},
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create snapshot: %w", err)
	}

	return snapshot, nil
}

// makeHTTPListener creates a listener on port 80 with HTTP connection manager
func makeHTTPListener() (*listener.Listener, error) {
	// Create HTTP connection manager config
	manager := &hcm.HttpConnectionManager{
		CodecType:  hcm.HttpConnectionManager_AUTO,
		StatPrefix: "http",
		RouteSpecifier: &hcm.HttpConnectionManager_Rds{
			Rds: &hcm.Rds{
				ConfigSource: &core.ConfigSource{
					ResourceApiVersion: core.ApiVersion_V3,
					ConfigSourceSpecifier: &core.ConfigSource_Ads{
						Ads: &core.AggregatedConfigSource{},
					},
				},
				RouteConfigName: "http_routes",
			},
		},
		HttpFilters: []*hcm.HttpFilter{
			{
				Name: wellknown.Router,
				ConfigType: &hcm.HttpFilter_TypedConfig{
					TypedConfig: mustMarshalAny(&router.Router{}),
				},
			},
		},
	}

	// Marshal to Any
	pbst, err := anypb.New(manager)
	if err != nil {
		return nil, err
	}

	return &listener.Listener{
		Name: "http_listener",
		Address: &core.Address{
			Address: &core.Address_SocketAddress{
				SocketAddress: &core.SocketAddress{
					Protocol: core.SocketAddress_TCP,
					Address:  "0.0.0.0",
					PortSpecifier: &core.SocketAddress_PortValue{
						PortValue: 80,
					},
				},
			},
		},
		FilterChains: []*listener.FilterChain{
			{
				Filters: []*listener.Filter{
					{
						Name: wellknown.HTTPConnectionManager,
						ConfigType: &listener.Filter_TypedConfig{
							TypedConfig: pbst,
						},
					},
				},
			},
		},
	}, nil
}

// makeEmptyRouteConfig creates a route configuration that returns 404 for all requests
func makeEmptyRouteConfig() *route.RouteConfiguration {
	return &route.RouteConfiguration{
		Name: "http_routes",
		VirtualHosts: []*route.VirtualHost{
			{
				Name:    "default_backend",
				Domains: []string{"*"},
				Routes: []*route.Route{
					{
						Match: &route.RouteMatch{
							PathSpecifier: &route.RouteMatch_Prefix{
								Prefix: "/",
							},
						},
						Action: &route.Route_DirectResponse{
							DirectResponse: &route.DirectResponseAction{
								Status: 404,
								Body: &core.DataSource{
									Specifier: &core.DataSource_InlineString{
										InlineString: "No apps exposed\n",
									},
								},
							},
						},
					},
				},
			},
		},
	}
}

// makeCluster creates a cluster for an app service
func makeCluster(name string, host string, port uint32) *cluster.Cluster {
	return &cluster.Cluster{
		Name:                 name,
		ConnectTimeout:       durationpb.New(5 * 1000000000), // 5 seconds in nanoseconds
		ClusterDiscoveryType: &cluster.Cluster_Type{Type: cluster.Cluster_STRICT_DNS},
		LbPolicy:             cluster.Cluster_ROUND_ROBIN,
		LoadAssignment: &endpoint.ClusterLoadAssignment{
			ClusterName: name,
			Endpoints: []*endpoint.LocalityLbEndpoints{
				{
					LbEndpoints: []*endpoint.LbEndpoint{
						{
							HostIdentifier: &endpoint.LbEndpoint_Endpoint{
								Endpoint: &endpoint.Endpoint{
									Address: &core.Address{
										Address: &core.Address_SocketAddress{
											SocketAddress: &core.SocketAddress{
												Protocol: core.SocketAddress_TCP,
												Address:  host,
												PortSpecifier: &core.SocketAddress_PortValue{
													PortValue: port,
												},
											},
										},
									},
								},
							},
						},
					},
				},
			},
		},
	}
}

// mustMarshalAny marshals a protobuf message to Any, panicking on error
func mustMarshalAny(msg proto.Message) *anypb.Any {
	a, err := anypb.New(msg)
	if err != nil {
		panic(fmt.Sprintf("failed to marshal %T to Any: %v", msg, err))
	}
	if a == nil {
		panic(fmt.Sprintf("anypb.New returned nil for %T", msg))
	}
	if a.TypeUrl == "" {
		panic(fmt.Sprintf("anypb.New returned empty TypeUrl for %T", msg))
	}
	return a
}

// Exposure represents a service exposure (minimal interface to avoid import cycle)
type Exposure struct {
	ID            string
	AppName       string
	Protocol      string
	Hostname      string
	ContainerPort uint32
	HostPort      uint32
}

// BuildSnapshotFromExposures creates a snapshot from a list of exposures
func BuildSnapshotFromExposures(version string, exposures []*Exposure) (*cache.Snapshot, error) {
	var listeners []types.Resource
	var routes []types.Resource
	var clusters []types.Resource

	// Separate HTTP and TCP exposures
	var httpExposures []*Exposure
	var tcpExposures []*Exposure

	for _, exp := range exposures {
		if exp.Protocol == "http" {
			httpExposures = append(httpExposures, exp)
		} else if exp.Protocol == "tcp" {
			tcpExposures = append(tcpExposures, exp)
		}
	}

	// Build HTTP listener and routes if we have HTTP exposures
	if len(httpExposures) > 0 {
		httpListener, err := makeHTTPListener()
		if err != nil {
			return nil, fmt.Errorf("failed to create HTTP listener: %w", err)
		}
		listeners = append(listeners, httpListener)

		// Build route config with all HTTP exposures
		routeConfig := makeRouteConfigFromExposures(httpExposures)
		routes = append(routes, routeConfig)

		// Build clusters for HTTP exposures
		for _, exp := range httpExposures {
			clusterName := fmt.Sprintf("cluster_%s", exp.ID)
			cluster := makeCluster(clusterName, exp.AppName, exp.ContainerPort)
			clusters = append(clusters, cluster)
		}
	} else {
		// No HTTP exposures, use empty route config
		httpListener, err := makeHTTPListener()
		if err != nil {
			return nil, fmt.Errorf("failed to create HTTP listener: %w", err)
		}
		listeners = append(listeners, httpListener)
		routes = append(routes, makeEmptyRouteConfig())
	}

	// Build TCP listeners for TCP exposures
	for _, exp := range tcpExposures {
		tcpListener, err := makeTCPListener(exp.ID, exp.HostPort, exp.AppName, exp.ContainerPort)
		if err != nil {
			return nil, fmt.Errorf("failed to create TCP listener for %s: %w", exp.ID, err)
		}
		listeners = append(listeners, tcpListener)

		clusterName := fmt.Sprintf("cluster_%s", exp.ID)
		cluster := makeCluster(clusterName, exp.AppName, exp.ContainerPort)
		clusters = append(clusters, cluster)
	}

	// Build snapshot
	snapshot, err := cache.NewSnapshot(
		version,
		map[resource.Type][]types.Resource{
			resource.ClusterType:  clusters,
			resource.RouteType:    routes,
			resource.ListenerType: listeners,
		},
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create snapshot: %w", err)
	}

	return snapshot, nil
}

// makeRouteConfigFromExposures creates a route configuration from HTTP exposures
func makeRouteConfigFromExposures(exposures []*Exposure) *route.RouteConfiguration {
	virtualHosts := make([]*route.VirtualHost, 0, len(exposures))

	for _, exp := range exposures {
		clusterName := fmt.Sprintf("cluster_%s", exp.ID)

		// Match both hostname and hostname.local for mDNS compatibility
		domains := []string{exp.Hostname}
		if !strings.HasSuffix(exp.Hostname, ".local") {
			domains = append(domains, exp.Hostname+".local")
		}

		virtualHost := &route.VirtualHost{
			Name:    exp.Hostname,
			Domains: domains,
			Routes: []*route.Route{
				{
					Match: &route.RouteMatch{
						PathSpecifier: &route.RouteMatch_Prefix{
							Prefix: "/",
						},
					},
					Action: &route.Route_Route{
						Route: &route.RouteAction{
							ClusterSpecifier: &route.RouteAction_Cluster{
								Cluster: clusterName,
							},
						},
					},
				},
			},
		}
		virtualHosts = append(virtualHosts, virtualHost)
	}

	return &route.RouteConfiguration{
		Name:         "http_routes",
		VirtualHosts: virtualHosts,
	}
}

// makeTCPListener creates a TCP listener for a specific port
func makeTCPListener(id string, hostPort uint32, targetHost string, targetPort uint32) (*listener.Listener, error) {
	clusterName := fmt.Sprintf("cluster_%s", id)

	tcpProxy := &tcpproxy.TcpProxy{
		StatPrefix: fmt.Sprintf("tcp_%s", id),
		ClusterSpecifier: &tcpproxy.TcpProxy_Cluster{
			Cluster: clusterName,
		},
	}

	pbst, err := anypb.New(tcpProxy)
	if err != nil {
		return nil, err
	}

	return &listener.Listener{
		Name: fmt.Sprintf("tcp_listener_%s", id),
		Address: &core.Address{
			Address: &core.Address_SocketAddress{
				SocketAddress: &core.SocketAddress{
					Protocol: core.SocketAddress_TCP,
					Address:  "0.0.0.0",
					PortSpecifier: &core.SocketAddress_PortValue{
						PortValue: hostPort,
					},
				},
			},
		},
		FilterChains: []*listener.FilterChain{
			{
				Filters: []*listener.Filter{
					{
						Name: wellknown.TCPProxy,
						ConfigType: &listener.Filter_TypedConfig{
							TypedConfig: pbst,
						},
					},
				},
			},
		},
	}, nil
}

File: src/zeropoint_agent/xds/snapshot.py
"""Build Envoy xDS snapshots from a list of resolved Endpoints.

Mirrors internal/xds/snapshot.go from the Go agent. Each call returns
a fresh (listeners, routes, clusters) triple at a monotonic version.

HTTP endpoints share one Listener on :80 with HCM/RDS; each gets a
VirtualHost (domain = name and name.local) routed to its own Cluster.
TCP endpoints get one Listener per endpoint bound to host_port, with a
TcpProxy filter pointing at its Cluster. Clusters are STRICT_DNS with
a single endpoint at `<container>-main:<container_port>`.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import timedelta
from typing import List

import betterproto2
from envoy_data_plane.envoy.config.cluster.v3 import (
    Cluster, ClusterDiscoveryType, ClusterLbPolicy,
)
from envoy_data_plane.envoy.config.core.v3 import (
    Address, AggregatedConfigSource, ApiVersion, ConfigSource,
    SocketAddress, SocketAddressProtocol,
)
from envoy_data_plane.envoy.config.endpoint.v3 import (
    ClusterLoadAssignment, Endpoint as EpEndpoint, LbEndpoint,
    LocalityLbEndpoints,
)
from envoy_data_plane.envoy.config.listener.v3 import (
    Filter, FilterChain, Listener,
)
from envoy_data_plane.envoy.config.route.v3 import (
    Route, RouteAction, RouteConfiguration, RouteMatch, VirtualHost,
)
from envoy_data_plane.envoy.extensions.filters.http.router.v3 import Router
from envoy_data_plane.envoy.extensions.filters.network.http_connection_manager.v3 import (
    HttpConnectionManager, HttpConnectionManagerCodecType,
    HttpFilter, Rds,
)
from envoy_data_plane.envoy.extensions.filters.network.tcp_proxy.v3 import TcpProxy
from envoy_data_plane.google.protobuf import Any as PbAny

HTTP_LISTENER_NAME = "http_listener"
HTTP_ROUTES_NAME = "http_routes"
HTTP_LISTEN_PORT = 80


@dataclass
class ResolvedEndpoint:
    """An Endpoint plus the resolved facts needed to build a snapshot."""
    id: str                # graph id, stable resource discriminator
    name: str              # mDNS hostname / vhost name (http) or label (tcp)
    protocol: str          # "http" | "tcp"
    container: str         # docker DNS name on zeropoint-network (e.g. "echo-main")
    container_port: int    # port inside the container
    host_port: int         # only used for tcp


def _pack_any(msg: betterproto2.Message, type_url: str) -> PbAny:
    return PbAny(type_url=type_url, value=bytes(msg))


def _socket_address(host: str, port: int) -> Address:
    return Address(
        socket_address=SocketAddress(
            protocol=SocketAddressProtocol.TCP,
            address=host,
            port_value=port,
        )
    )


def _ads_config() -> ConfigSource:
    """ADS-pointing ConfigSource used by RDS inside the HTTP listener."""
    return ConfigSource(
        resource_api_version=ApiVersion.V3,
        ads=AggregatedConfigSource(),
    )


def make_http_listener() -> Listener:
    router = Router()
    hcm = HttpConnectionManager(
        codec_type=HttpConnectionManagerCodecType.AUTO,
        stat_prefix="http",
        request_timeout=timedelta(seconds=0),
        stream_idle_timeout=timedelta(seconds=600),
        request_headers_timeout=timedelta(seconds=300),
        use_remote_address=True,
        skip_xff_append=False,
        rds=Rds(
            config_source=_ads_config(),
            route_config_name=HTTP_ROUTES_NAME,
        ),
        http_filters=[
            HttpFilter(
                name="envoy.filters.http.router",
                typed_config=_pack_any(
                    router,
                    "type.googleapis.com/envoy.extensions.filters.http.router.v3.Router",
                ),
            ),
        ],
    )
    return Listener(
        name=HTTP_LISTENER_NAME,
        address=_socket_address("0.0.0.0", HTTP_LISTEN_PORT),
        filter_chains=[
            FilterChain(
                filters=[
                    Filter(
                        name="envoy.filters.network.http_connection_manager",
                        typed_config=_pack_any(
                            hcm,
                            "type.googleapis.com/envoy.extensions.filters.network."
                            "http_connection_manager.v3.HttpConnectionManager",
                        ),
                    ),
                ],
            ),
        ],
    )


def make_tcp_listener(ep: ResolvedEndpoint) -> Listener:
    cluster_name = _cluster_name(ep)
    proxy = TcpProxy(
        stat_prefix=f"tcp_{ep.id}",
        cluster=cluster_name,
    )
    return Listener(
        name=f"tcp_listener_{ep.id}",
        address=_socket_address("0.0.0.0", ep.host_port),
        filter_chains=[
            FilterChain(
                filters=[
                    Filter(
                        name="envoy.filters.network.tcp_proxy",
                        typed_config=_pack_any(
                            proxy,
                            "type.googleapis.com/envoy.extensions.filters."
                            "network.tcp_proxy.v3.TcpProxy",
                        ),
                    ),
                ],
            ),
        ],
    )


def _cluster_name(ep: ResolvedEndpoint) -> str:
    return f"cluster_{ep.id.replace('/', '_')}"


def make_cluster(ep: ResolvedEndpoint) -> Cluster:
    name = _cluster_name(ep)
    return Cluster(
        name=name,
        connect_timeout=timedelta(seconds=5),
        type=ClusterDiscoveryType.STRICT_DNS,
        lb_policy=ClusterLbPolicy.ROUND_ROBIN,
        load_assignment=ClusterLoadAssignment(
            cluster_name=name,
            endpoints=[
                LocalityLbEndpoints(
                    lb_endpoints=[
                        LbEndpoint(
                            endpoint=EpEndpoint(
                                address=_socket_address(ep.container, ep.container_port),
                            ),
                        ),
                    ],
                ),
            ],
        ),
    )


def make_route_config(http_endpoints: List[ResolvedEndpoint]) -> RouteConfiguration:
    vhosts: List[VirtualHost] = []
    if not http_endpoints:
        vhosts.append(VirtualHost(
            name="default_404",
            domains=["*"],
            routes=[Route(
                match=RouteMatch(prefix="/"),
                route=RouteAction(cluster="__unused__"),
            )],
        ))
    for ep in http_endpoints:
        domains = [ep.name]
        if not ep.name.endswith(".local"):
            domains.append(f"{ep.name}.local")
        vhosts.append(VirtualHost(
            name=ep.name,
            domains=domains,
            routes=[Route(
                match=RouteMatch(prefix="/"),
                route=RouteAction(
                    cluster=_cluster_name(ep),
                    timeout=timedelta(seconds=0),
                    idle_timeout=timedelta(seconds=300),
                ),
            )],
        ))
    return RouteConfiguration(name=HTTP_ROUTES_NAME, virtual_hosts=vhosts)


@dataclass
class Snapshot:
    """A self-contained xDS snapshot keyed by monotonic version."""
    version: str
    listeners: List[Listener]
    routes: List[RouteConfiguration]
    clusters: List[Cluster]


def build_snapshot(version: str, endpoints: List[ResolvedEndpoint]) -> Snapshot:
    http = [e for e in endpoints if e.protocol == "http"]
    tcp = [e for e in endpoints if e.protocol == "tcp"]

    listeners: List[Listener] = [make_http_listener()]
    for ep in tcp:
        listeners.append(make_tcp_listener(ep))

    routes = [make_route_config(http)]
    clusters = [make_cluster(e) for e in (http + tcp)]

    return Snapshot(
        version=version,
        listeners=listeners,
        routes=routes,
        clusters=clusters,
    )
